Upgraded to Table Schema spec v1 null/types/constraints semantics

roll opened the pull request:
- fixes #150 
- fixes #94

---

This PR makes `tableschema-py` to act as it described in Table Schema v1 spec. This release of the spec has allowed to greatly simplify dataflow inside the library. Now value processing contains three dedicated steps:
- test for null value (based on schema level `missingValues`)
- cast value to type (based on pure cast functions for every Table Schema type - we support 100% of types/formats)
- check value against constraints (based on pure check functions for every Table Schema constraint)

Also cast tests now is much more readable so it could be reviewed even by non-tech people.

For review (diff is kinda verbose because of legacy stuff):
 - https://github.com/frictionlessdata/tableschema-py/tree/null-types-constraints-v1
- https://github.com/frictionlessdata/tableschema-py/blob/null-types-constraints-v1/tableschema/field.py (main dataflow)

pwalsh commented:
Also note there is a possible change in behaviour when casting. 

Previously, if a format was invalid or simply not handled in the current implementation, then the default format would be used:

- https://github.com/frictionlessdata/tableschema-py/blob/master/tableschema/types/base.py#L119

Now, there is no default casting handler (as casting is not based on inheritance), and `default` is not consistently a fallback in each casting function (for example [here](https://github.com/frictionlessdata/tableschema-py/blob/null-types-constraints-v1/tableschema/types/geopoint.py#L15) an unknown format keyword will eventually raise an error, yet [here](https://github.com/frictionlessdata/tableschema-py/blob/null-types-constraints-v1/tableschema/types/string.py#L17) `default`, or any other unknown keyword will pass through, consistent with the previous implementation). Am I missing something in the control flow?

roll commented:
Thanks.

> All cast functions return strings for errors. It means it is "impossible" to use the cast functions alone, as exception handling now requires a strict dependency on the Field class

Cast functions is not a part of public API.

> There is now a "legacy" package which is still in use? How is it still required, considering the the above refactor

Yes. It's used by `infer` function. Haven't touched it for now per https://github.com/frictionlessdata/tableschema-py/issues/105#issuecomment-288161154

> We now use six (another refactoring?), so do we still need the simpler compat module?

`six` is a standard accros all our python code so it has been used by default for all newly written `isinstance(value, six.string_types)` (see below in type related changes section). `compat` is still in use in some place - will be handled in #83. That's just not a focus of this PR.   

> I can't see where you cast constraint values, so for example if someone had constraints like '3' or '2017-01-01' these will fail, as your new check functions work on the assumption of values already cast. Or, have I missed something in here?

It's covered here - https://github.com/frictionlessdata/tableschema-py/blob/null-types-constraints-v1/tableschema/field.py#L153-L167 (and cast function preparation above). `Table/Schema/Field` passes all pre-PR tests not related to spec-v1 changes.

That's change is really important because existent code was making `goodtables.py` and `GTIO` to work really ineffective. For example for table with 20 columns and 10 000 rows this code `Field.__prepare_cast_function` and `Field.__prepare_checks_functions` (including constraint casting) was called for every `Field.cast_value` call means 200 000 calls.

> I don't see any changes made to foreign key support. That may be ok, but please confirm.

That's correct. Per https://github.com/frictionlessdata/tableschema-py/issues/33#issuecomment-288160861 it will be covered later - #33. On `Schema` level it's already compliant to spec-v1 after #157.

> Previously, if a format was invalid or simply not handled in the current implementation, then the default format would be used. Now, there is no default casting handler (as casting is not based on inheritance), and default is not consistently a fallback in each casting function (for example here an unknown format keyword will eventually raise an error, yet here default, or any other unknown keyword will pass through, consistent with the previous implementation). Am I missing something in the control flow?

Now it's an invariant provided by spec - https://specs.frictionlessdata.io/schemas/table-schema.json - valid descriptor can't have not-supported `format`.

---

Changes related to field in this PR:
- `missingValues` moved from field level to schema level and set to `['']`
- it's hard to say what's exactly changed to null/types/constraints implementation because even with my knowledge of codebase I was hardly understanding what's really happening with value (esp. null values management) - now it's just compliant to the spec (including our latest discussion about constraints casting)
- removed duplication of constraints implementation (in `constraints` folder and `Field.__check_*`) which has been introduced because initial types system doesn't work for `goodtables-py` on granular constraint checks

---

Changes related to types in PR:
- general
  - all types when applicable wasn't compliant to spec with handling cast values of other types e.g. `IntegerField.cast_value(3.14) == 3` instead of `IntegerField.cast_value(3.14) == ERROR`. Now all types accept only string (lexical representation) or cast value of proper type.
  - types `duration`, `year` and `yearmonth` has been added by side contributor earlier (#152) so only changed after this PR is covered by this diff
- date
  - changed format=`fmt:<PATTERN>` to `format=<PATTERN>`
- datetime
  - changed format=`fmt:<PATTERN>` to `format=<PATTERN>`
- geojson
  - added support for `format=topojson` without validation (#158)
- number
  - now currency symbols will be removed only if there is `currency=True`
  - not changed in spec but it was not compliant:
    - fixed default `groupChar` to `''` from `','` (may be we're better change spec instead if it's really breaking to OS cc @akariv )
    - fixed handling values with percent symbols - now it will be divided by 100
- time
  - changed format=`fmt:<PATTERN>` to `format=<PATTERN>`
- year
  - now checked to have length == 4 as a string 
  - now checked to be inside [0, 9999] as an integer
  
---

> refactoring that is purely aesthetic (even taste driven - OO to functional).

Just can't agree on it - reducing complexity, improving testability and effectiveness is not an aesthetic changes. Esp. before porting this library to other 8 languages. Even I with all my knowledge of the codebase wasn't sure what's needed to be done to be sure that `Field` is null/types/constraints compliant to spec-v1 without simplifying things.

pwalsh commented:
Now it's an invariant provided by spec - https://specs.frictionlessdata.io/schemas/table-schema.json - valid descriptor can't have not-supported `format`.

Ok. So I think you'll have to add an extra test for `<PATTERN>` on date/time/datetime, as the spec in JSON Schema does not handle that in the format enum, so it must fail in that case. Could you also change the `enum` to `pattern` in the JSON Schema spec to support this (I don't think JSON Schema supports a pattern string within an enum).

pwalsh commented:
Please add support for `fmt:` on date/time/datetime fields, with a warning that this was removed in v1 spec

pwalsh commented:
Cast functions is not a part of public API.

```
from jsontableschema import types

is_array = types.ArrayType().cast
is_array('1')

# raises InvalidArrayType: Not an array
```

akariv commented:
So, two comments:
Overall I really enjoyed reviewing this PR - going over commit by commit really made this initially daunting PR into a very structured experience. Also, technically speaking, the result is clearly more organised and better designed than the original code.

I did expect, however, to see the changes made to the logic so that I could understand the change in functionality (and compare that to the changes in the spec). Instead, I saw new files added with that functionality added from scratch - which made it effectively impossible to do that sort of comparison.

The problem with such rewrites is mostly from a maturity point of view - the previous code had already matured and all sort of edge cases were handled (and in a certain way). When rewriting (if not done in a very specific way), the maturity of the code 'resets' and the entire process needs to start over.
For example, now I can't be sure that the same edge cases are handled as well, and that the decisions for each edge case remain the same. 
These subtle changes might introduce new bugs, and could cause a ripple effect of other things breaking in whatever code that's using these libraries.

2nd comment is regarding percent:
I think that the 'divide by 100' behaviour should be tossed away from the spec as well :) , but as long as we're keeping it, it should be limited to `%％﹪٪` only and not the other `‰‱` signs, which need to be divided by 1,000 and 10,000 respectively.

amercader commented on tableschema/types/array.py:
```diff
@@ -11,7 +11,7 @@
 # Module API
 
-def cast_array_default(value):
+def cast_array(format, value):
```
Not sure if there is a previous context on this but wouldn't it make more sense to move `format` to an optional keyword argument that defaults to `default` as this is what people will need 90% of the time: `cast_*(value)`. Others can use `cast_string(value, format='email')` or just `cast_string(value, 'email')`.

amercader commented on tableschema/types/array.py:
```diff
@@ -4,42 +4,21 @@
-        if isinstance(value, self.python_type):
-            return value
-
+def cast_array(format, value):
```
Not sure if there is a previous context on this but wouldn't it make more sense to move `format` to an optional keyword argument that defaults to `default` as this is what people will need 90% of the time: `cast_*(value)`. Others can use `cast_string(value, format='email')` or just `cast_string(value, 'email')`.

roll replied:
This interface is totally internal so I've selected this only to make tests - https://github.com/frictionlessdata/tableschema-py/blob/null-types-constraints-v1/tests/types/test_datetime.py - to look more like DSL (I mean invocation consistency - type -> format -> value). But for end-user there is no difference.

pwalsh replied:
@roll I'm not convinced it was a good idea to make casting internal. simple casting of values is really useful. Do you remember I wanted us to use https://github.com/pudo/typecast directly for this back then?

Whatever the case, I agree with @amercader that, as format is essentially optional (defaults to `default`, we don't require users to provide `default`), then a more conventional function signature would be with format as a keyword argument.

roll commented on tableschema/types/yearmonth.py:
```diff
@@ -4,44 +4,22 @@
-        if isinstance(value, self.python_type):
-            return value
-
+def cast_yearmonth(format, value):
```
Thanks, great catch - https://github.com/frictionlessdata/specs/issues/385#issuecomment-291427679

roll commented on tableschema/types/geojson.py:
```diff
@@ -4,66 +4,29 @@
+        except Exception:
+            return ERROR
+    elif format == 'topojson':
+        if not isinstance(value, dict):
```
Awesome. Logget - #158

roll commented:
@pwalsh 
cc @amercader 

Wanted to answer you after all tweaks but if you're already on topic.

Here is your example:

```
from jsontableschema import types

is_array = types.ArrayType().cast
is_array('1')

# raises InvalidArrayType: Not an array
```

Here is `Field API` example:

```
from jsontableschema import Field

is_array = types.Field({'type': 'array'}).cast
is_array('1')

# raises InvalidCastError: Not an array
```

For the user it works exactly the same.

Also I want to add additional information as attributes to `InvalidCastError` like type, format, value etc Instead of having this enterprise-Java-like big exceptions family. But I don't think it's a object of this PR. I'm tracking it here - https://github.com/frictionlessdata/tableschema-py/issues/80

roll commented:
@pwalsh 
> Please add support for fmt: on date/time/datetime fields, with a warning that this was removed in v1 spec
done 

@amercader 
`yearmonth` is fixed to be as written here - https://pre-v1.frictionlessdata.io/json-table-schema/#gyearmonth (`namedtuple` as python value). Also following our Slack discussion - `geopoint` also now `namedtuple` as python value.

@akariv 
Thanks. It's a great input. And I agree. This changes was a difficult decision. When I've stuck with nulls/constraints with not updated codebase just decided to go with solution I had around a 6 month in my head (#94). I've weighted all time/code pros and cons and realized that this will be the most optimal. We have a good test coverage (including high-level `Table/Schema/Field`) and I've just reuse edge-case tests on low-level.

roll commented:
@pwalsh 
It seems I've missed this one:
> Ok. So I think you'll have to add an extra test for <PATTERN> on date/time/datetime, as the spec in JSON Schema does not handle that in the format enum, so it must fail in that case. Could you also change the enum to pattern in the JSON Schema spec to support this (I don't think JSON Schema supports a pattern string within an enum).

Not sure what do you mean. It's not possible to validated strptime `<PATTERN>` so we just need to remove `enum` from `date/datetime/time` (https://specs.frictionlessdata.io/schemas/table-schema.json).

pwalsh commented:
@roll 

I don't see a need to keep the legacy package. The other changes are so extensive, and you are keeping the legacy package [basically for this function](https://github.com/frictionlessdata/tableschema-py/blob/null-types-constraints-v1/tableschema/infer.py#L151), and a couple of lines [in this method](https://github.com/frictionlessdata/tableschema-py/blob/null-types-constraints-v1/tableschema/infer.py#L104). 

This does not mean you need to rewrite the inference engine in this pull request. Issue #105 is clearly designated for that. However, considering the scope of other changes in this PR, Just updating those lines of code to work with Fields would allow you to remove this newly created legacy package.

roll commented:
@pwalsh 
will do

roll commented:
@pwalsh 
Good catch - it was easy to update `infer`. So I've removed legacy code.
